### DISCUSSION TITLE
mermaid theming: classDef requires htmlLabels: true

### DISCUSSION THREAD

**Discussion by dmi3kno** [quarto-dev/quarto-cli#6209] [Q&A] [answered]
### Description

I am very encouraged to see a new approach to [styling mermaid diagrams ](https://quarto.org/docs/authoring/diagrams.html#mermaid-theming) in Quarto 1.3! Thank you!

My question is related to using `classDef` property inside the diagram to style some particular nodes. It seem to be working for everything other than the color:

````qmd
---
title: "Untitled"
format: 
   html:
     mermaid: 
       theme: default
---

```{mermaid}
flowchart LR
  classDef dark color:#fff,fill:#0d5caa
  Lorem --> Ipsum --> Dolor
  class Lorem,Dolor dark
```

````

Github rendering
```mermaid
flowchart LR
  classDef dark color:#fff,fill:#0d5caa
  Lorem --> Ipsum --> Dolor
  class Lorem,Dolor dark
```

Quarto rendering

![Screenshot from 2023-07-14 12-10-38](https://github.com/quarto-dev/quarto-cli/assets/13419011/6c1f06ed-9ebb-41b3-bc2a-9b1a224044ab)

I came across mermaid-js/mermaid#1955 in mermaid, but I don't think it is related to that. What is more likely is that `classDef` color gets overwritten by the document CSS. Is that something you guys can confirm? What would be the proper line of attack here? How can I make sure `classDef` stipulated `color` takes precedence over `$body-color`?

**Comment by cscheid** [comment]
You'll have to inspect the CSS with the developer tools in your browser and investigate, I'm afraid. Each mermaid diagram uses its own set of CSS rules and we don't have the cycles to track all of it.

**Reply by dmi3kno** [reply, marked as answer]
Thank you for the tip. I should have inspected it before writing, of course. The difference seems to be that github is using a different HTML construction to render the flowchart-label:
- github: `<foreignObject><div><span>...</span> </div></foreignObject>` 
- Quarto: `<text><tspan>...</tspan></text>`

The Mermaid CSS is very similar, but since `<text>` has its own styling, it overrides the style assigned at the container group `<g>` (the custom `dark` class).

In order to make Quarto use the same HTML construct one should add 
```
%%{init:{"theme":"base", "flowchart": {"htmlLabels": true}}}%%
```

This options is not currently exposed in YAML, but I feel like it should be:

```qmd
---
title: "Untitled"
format: 
   html:
     mermaid: 
       theme: default
       flowchart:
         htmlLabels: true
---

```
This seems to be simiar to quarto-dev/quarto-cli#6064. Is there a cleaner way to force HTML labels by default?

**Comment by dmi3kno** [comment]
While we are on it, can I please suggest an improvement to documentation?

This [page](https://quarto.org/docs/authoring/diagrams.html#customizing-mermaid) suggests that the properties are set as
```
$mermaid-fg-color--lighter: $body-color-lighter;
$mermaid-fg-color--lightest: $body-color-lightest;
```
When if fact they are [set as](https://github.com/quarto-dev/quarto-cli/blob/ae8c20b27713e19fd21e692779083a1198e3462d/src/resources/formats/html/_quarto-rules.scss#L642)
```
$mermaid-fg-color--lighter: lighten($body-color, 10%) !default;
$mermaid-fg-color--lightest: lighten($body-color, 20%) !default;
```

I don't think that `body-color-lightest` or `body-color-lighter` are even defined. Would be good to clarify those for future users. What do you think @cderv ?

**Reply by cscheid** [reply]
Ouch, those definitely need changing. I think this was a copy-and-paste error.

**Reply by cderv** [reply]
Thanks @dmi3kno ! It has been fixed already in https://github.com/quarto-dev/quarto-web/commit/0e49089e4755c241851b2ebf01cb831bc9cf0d94  (Thanks Carlos ! )

Thanks for the report!
